File: code/packages/python/ecmascript-es3-parser/tests/test_ecmascript_es3_parser.py
"""Tests for the ECMAScript 3 (1999) Parser.

ES3 adds try/catch/finally/throw, strict equality, instanceof, and regex
to the ES1 grammar.
"""

from __future__ import annotations

from lang_parser import ASTNode
from lexer import Token

from ecmascript_es3_parser import create_es3_parser, parse_es3


def find_nodes(node: ASTNode, rule_name: str) -> list[ASTNode]:
    results: list[ASTNode] = []
    if node.rule_name == rule_name:
        results.append(node)
    for child in node.children:
        if isinstance(child, ASTNode):
            results.extend(find_nodes(child, rule_name))
    return results


def find_tokens(node: ASTNode) -> list[Token]:
    tokens: list[Token] = []
    for child in node.children:
        if isinstance(child, Token):
            tokens.append(child)
        elif isinstance(child, ASTNode):
            tokens.extend(find_tokens(child))
    return tokens


# ============================================================================
# Test: ES3 try/catch/finally (NEW)
# ============================================================================


class TestTryCatchFinally:

    def test_try_catch(self) -> None:
        ast = parse_es3("try { var x = 1; } catch (e) { var y = 2; }")
        try_stmts = find_nodes(ast, "try_statement")
        assert len(try_stmts) == 1

    def test_try_finally(self) -> None:
        ast = parse_es3("try { var x = 1; } finally { var y = 2; }")
        try_stmts = find_nodes(ast, "try_statement")
        assert len(try_stmts) == 1

    def test_try_catch_finally(self) -> None:
        ast = parse_es3("try { } catch (e) { } finally { }")
        try_stmts = find_nodes(ast, "try_statement")
        assert len(try_stmts) == 1

    def test_throw(self) -> None:
        ast = parse_es3('throw 1;')
        throw_stmts = find_nodes(ast, "throw_statement")
        assert len(throw_stmts) == 1


# ============================================================================
# Test: ES3 Strict Equality in Expressions
# ============================================================================


class TestStrictEqualityParsing:

    def test_strict_equals_expression(self) -> None:
        ast = parse_es3("var x = 1 === 2;")
        eq_nodes = find_nodes(ast, "equality_expression")
        assert len(eq_nodes) >= 1


# ============================================================================
# Test: ES1 Features Still Parse in ES3
# ============================================================================


class TestES1Compatibility:

    def test_var_declaration(self) -> None:
        ast = parse_es3("var x = 1;")
        assert ast.rule_name == "program"

    def test_function_declaration(self) -> None:
        ast = parse_es3("function foo() { }")
        func_decls = find_nodes(ast, "function_declaration")
        assert len(func_decls) == 1

    def test_if_else(self) -> None:
        ast = parse_es3("if (x) { } else { }")
        if_stmts = find_nodes(ast, "if_statement")
        assert len(if_stmts) == 1

    def test_while_loop(self) -> None:
        ast = parse_es3("while (x) { }")
        while_stmts = find_nodes(ast, "while_statement")
        assert len(while_stmts) == 1

    def test_for_loop(self) -> None:
        ast = parse_es3("for (var i = 0; i; i) { }")
        for_stmts = find_nodes(ast, "for_statement")
        assert len(for_stmts) == 1

    def test_switch(self) -> None:
        ast = parse_es3("switch (x) { case 1: var y = 2; }")
        switch_stmts = find_nodes(ast, "switch_statement")
        assert len(switch_stmts) == 1

    def test_object_literal(self) -> None:
        ast = parse_es3("var x = { a: 1 };")
        obj_nodes = find_nodes(ast, "object_literal")
        assert len(obj_nodes) == 1


# ============================================================================
# Test: Factory Function
# ============================================================================


class TestCreateES3Parser:

    def test_creates_parser(self) -> None:
        parser = create_es3_parser("var x = 1;")
        assert hasattr(parser, "parse")

    def test_factory_produces_same_result(self) -> None:
        source = "try { } catch (e) { }"
        ast_direct = parse_es3(source)
        ast_factory = create_es3_parser(source).parse()
        assert ast_direct.rule_name == ast_factory.rule_name

File: code/packages/ruby/ecmascript_es3_parser/test/test_ecmascript_es3_parser.rb
# frozen_string_literal: true

require_relative "test_helper"

# ================================================================
# Tests for the ECMAScript 3 (ES3) Parser
# ================================================================
#
# These tests verify ES3-specific grammar features:
#   - try/catch/finally/throw statements
#   - Strict equality (===, !==) in expressions
#   - instanceof in relational expressions
#   - All ES1 features still work
# ================================================================

class TestEcmascriptEs3Parser < Minitest::Test
  ASTNode = CodingAdventures::Parser::ASTNode
  TT = CodingAdventures::Lexer::TokenType

  def parse(source)
    CodingAdventures::EcmascriptEs3Parser.parse(source)
  end

  # ------------------------------------------------------------------
  # Basic: var declaration (inherited from ES1)
  # ------------------------------------------------------------------

  def test_var_declaration
    ast = parse("var x = 1 + 2;")

    assert_equal "program", ast.rule_name
    source_el = ast.children.find { |c| c.is_a?(ASTNode) }
    refute_nil source_el
  end

  # ------------------------------------------------------------------
  # ES3-specific: try/catch statement
  # ------------------------------------------------------------------

  def test_try_catch
    ast = parse("try { var x = 1; } catch (e) { var y = 2; }")

    assert_equal "program", ast.rule_name
    all_tokens = collect_tokens(ast)
    try_kw = all_tokens.find { |t| t.type == TT::KEYWORD && t.value == "try" }
    refute_nil try_kw, "Expected 'try' keyword in AST"
    catch_kw = all_tokens.find { |t| t.type == TT::KEYWORD && t.value == "catch" }
    refute_nil catch_kw, "Expected 'catch' keyword in AST"
  end

  # ------------------------------------------------------------------
  # ES3-specific: try/finally statement
  # ------------------------------------------------------------------

  def test_try_finally
    ast = parse("try { var x = 1; } finally { var z = 3; }")

    all_tokens = collect_tokens(ast)
    finally_kw = all_tokens.find { |t| t.type == TT::KEYWORD && t.value == "finally" }
    refute_nil finally_kw, "Expected 'finally' keyword in AST"
  end

  # ------------------------------------------------------------------
  # ES3-specific: throw statement
  # ------------------------------------------------------------------

  def test_throw_statement
    ast = parse("throw x;")

    all_tokens = collect_tokens(ast)
    throw_kw = all_tokens.find { |t| t.type == TT::KEYWORD && t.value == "throw" }
    refute_nil throw_kw, "Expected 'throw' keyword in AST"
  end

  # ------------------------------------------------------------------
  # Expression statement (inherited from ES1)
  # ------------------------------------------------------------------

  def test_expression_statement
    ast = parse("1 + 2;")

    assert_equal "program", ast.rule_name
    source_el = ast.children.find { |c| c.is_a?(ASTNode) }
    refute_nil source_el
  end

  # ------------------------------------------------------------------
  # Operator precedence
  # ------------------------------------------------------------------

  def test_operator_precedence
    ast = parse("1 + 2 * 3;")

    all_tokens = collect_tokens(ast)
    assert all_tokens.any? { |t| t.type == TT::PLUS }
    assert all_tokens.any? { |t| t.type == TT::STAR }
  end

  # ------------------------------------------------------------------
  # Multiple statements
  # ------------------------------------------------------------------

  def test_multiple_statements
    ast = parse("var x = 1;var y = 2;")

    assert_equal "program", ast.rule_name
    source_elements = ast.children.select { |c| c.is_a?(ASTNode) && c.rule_name == "source_element" }
    assert_equal 2, source_elements.length
  end

  # ------------------------------------------------------------------
  # Root is always 'program'
  # ------------------------------------------------------------------

  def test_root_is_program
    ast = parse("1;")
    assert_equal "program", ast.rule_name
  end

  # ------------------------------------------------------------------
  # Empty program
  # ------------------------------------------------------------------

  def test_empty_program
    ast = parse("")
    assert_equal "program", ast.rule_name
  end

  # ------------------------------------------------------------------
  # Function declaration
  # ------------------------------------------------------------------

  def test_function_declaration
    ast = parse("function foo(a) { return a; }")

    all_tokens = collect_tokens(ast)
    fn_kw = all_tokens.find { |t| t.type == TT::KEYWORD && t.value == "function" }
    refute_nil fn_kw
  end

  # ------------------------------------------------------------------
  # Grammar path resolution
  # ------------------------------------------------------------------

  def test_grammar_path_exists
    assert File.exist?(CodingAdventures::EcmascriptEs3Parser::ES3_GRAMMAR_PATH),
      "es3.grammar file should exist at #{CodingAdventures::EcmascriptEs3Parser::ES3_GRAMMAR_PATH}"
  end

  private

  def collect_tokens(node)
    tokens = []
    return tokens unless node.is_a?(ASTNode)

    node.children.each do |child|
      if child.is_a?(CodingAdventures::Lexer::Token)
        tokens << child
      elsif child.is_a?(ASTNode)
        tokens.concat(collect_tokens(child))
      end
    end
    tokens
  end
end
